notebook/scale-test.py

```diff
@@ -26,86 +26,88 @@ def get_cookie(session, name):
     headers = service_auth_headers(deploy_config, 'workshop', authorize_target=False)
 
     async with aiohttp.ClientSession(raise_for_status=True) as session:
-        # make sure notebook is up
-        async with session.get(
-                deploy_config.url('workshop', ''),
-                headers=headers) as resp:
-            await resp.text()
-
-        log.info(f'{i} loaded notebook home page')
-
-        # log in as workshop guest
-        # get csrf token
-        async with session.get(
-                deploy_config.url('workshop', '/login'),
-                headers=headers) as resp:
-            pass
-
-        data = aiohttp.FormData()
-        data.add_field(name='name', value=args.workshop)
-        data.add_field(name='password', value=args.password)
-        data.add_field(name='_csrf', value=get_cookie(session, '_csrf'))
-        async with session.post(
-                deploy_config.url('workshop', '/login'),
-                data=data,
-                headers=headers) as resp:
-            pass
-
-        log.info(f'{i} logged in')
-
-        # create notebook
-        # get csrf token
-        async with session.get(
-                deploy_config.url('workshop', '/notebook'),
-                headers=headers) as resp:
-            pass
-
-        data = aiohttp.FormData()
-        data.add_field(name='_csrf', value=get_cookie(session, '_csrf'))
-        async with session.post(
-                deploy_config.url('workshop', '/notebook'),
-                data=data,
-                headers=headers) as resp:
-            pass
-
-        log.info(f'{i} created notebook')
-
-        start = time.time()
-
-        # wait for notebook ready
-        ready = False
-        attempt = 0
-        # 5 attempts overkill, should only take 2: Scheduling => Running => Ready
-        while not ready and attempt < 5:
-            async with session.ws_connect(
-                    deploy_config.url('workshop', '/notebook/wait', base_scheme='ws'),
-                    headers=headers) as ws:
-                async for msg in ws:
-                    if msg.data == '1':
-                        ready = True
-            attempt += 1
-
-        end = time.time()
-        duration = end - start
-
-        log.info(f'{i} notebook state {ready} duration {duration}')
-
-        # delete notebook
-        # get csrf token
-        async with session.get(
-                deploy_config.url('workshop', '/notebook'),
-                headers=headers) as resp:
-            pass
-
-        data = aiohttp.FormData()
-        data.add_field(name='_csrf', value=get_cookie(session, '_csrf'))
-        async with session.post(
-                deploy_config.url('workshop', '/notebook/delete'),
-                data=data,
-                headers=headers) as resp:
-            pass
-
-        log.info(f'{i} notebook delete, done.')
+        try:
```

Comment:
You need to set duration = None, ready = False before the try or you'll get an undefined variable error in the case of exception.

Reply:
I don't suppress the exception, I just make sure we delete the notebook.

Do you want to catch the exception and return None, None?

Reply:
That would also work.  You return after the try ... except block and set those variables that are set only in the body of the try ... except.

Reply:
Oh, oops, I see what you are saying.

Yes, we should supress the exception and return None, None.  A not true ready is treated as an error by main.  Either this, or it needs to be modified to handle the exceptions generated in teh children of the gather.

Reply:
To clarify further, this fixes the cleanup, but it doesn't treat exceptions in the made code as failed notebook attempts.  Can you make this change also?